lib/puppet/pal/pal_impl.rb

```diff
@@ -168,29 +168,26 @@ def self.with_catalog_compiler(
       # Use the manifest setting
       manifest_file = Puppet[:manifest]
     else
-      # An "undef" code_string is the only way to override Puppet[:manifest] & Puppet[:code] settings since an
-      # empty string is taken as Puppet[:code] not being set.
+      # An "undef" code_string is the only way to override Puppet[:manifest] setting & Puppet.lookup(:code)
+      # since an empty string is taken as Puppet.lookup(:code) not being set.
       #
       if manifest_file.nil? && code_string.nil?
         code_string = 'undef'
       end
     end
 
-    # We need to make sure to set these back when we're done
-    previous_tasks_value = Puppet[:tasks]
-    previous_code_value = Puppet[:code]
+    old_code = Puppet.lookup(:code)
+    Puppet.push_context({code: code_string}) unless code_string.nil?
+    Puppet.override({tasks: false}) do
+      # After the assertions, if code_string is non nil - it has the highest precedence
 
-    Puppet[:tasks] = false
-    # After the assertions, if code_string is non nil - it has the highest precedence
-    Puppet[:code] = code_string unless code_string.nil?
-
-    # If manifest_file is nil, the #main method will use the env configured manifest
-    # to do things in the block while a Script Compiler is in effect
-    main(manifest_file, facts, variables, :catalog, &block)
+      # If manifest_file is nil, the #main method will use the env configured manifest
+      # to do things in the block while a Script Compiler is in effect
+      main(manifest_file, facts, variables, :catalog, &block)
+    end
   ensure
     # Clean up after ourselves
-    Puppet[:tasks] = previous_tasks_value
-    Puppet[:code] = previous_code_value
+    Puppet.push_context({code: old_code})
```

Comment:
I would think you'd be popping instead of pushing here to get back to previous state, or use override instead of this.

Reply:
Yeah, this is dumb.

Reply:
yeah, not sure why you wouldn't override both `tasks` and `code` at once